sound/soc/sof/trace.c

```diff
@@ -95,8 +109,21 @@ static ssize_t sof_dfsentry_trace_read(struct file *file, char __user *buffer,
 	return count;
 }
 
+static int sof_dfsentry_trace_open(struct inode *inode, struct file *file)
+{
+	struct snd_sof_dfsentry *dfse = inode->i_private;
+	struct snd_sof_dev *sdev = dfse->sdev;
+
+	if (inode->i_private)
+		file->private_data = inode->i_private;
+	sdev->dtrace_eof = false;
+	sdev->host_offset = 0;
```

Comment:
@kv2019i should the host_offset to reset too?

Reply:
@ranj063 Uhm, yes, but I do do reset "sdev->host_offset" here, right? :) Or did you mean the reverse that it should _not_ be reset here?

If you refer to the file node position, that's something we don't really have much control over, which makes this a lot more complicated in general. The sof-logger implementation does a lot of seeks, so we need to keep supporting arbitrary seeks (although sof-logger's seeks are sequential, but it still does seeks). So our debugfs read() just has to work with whatever position the client wants to read from.

My patch now covers this for most part. One major downside is that traces disappear when runtime resume is done. So if sof-logger doesn't read the traces in time, then you lose those. But with typical timeouts for runtime-pm, sof-logger has enough time to get the traces read out. 

And of course, this does not help to get late traces between suspend-resume. But you'll anyways have a blackout period when the trace is not set up so FW cannot push the traces anywhere.

Reply:
@kv2019i sorry I meant the hsot_offset should not be reset here. we reset the host_offset when we resume from suspend. Could this possibly why you're losing the trace from before the suspend?

Reply:
@ranj063 I tested without, but we hit one corner case if we don't reset at open:

- FW is in runtime suspend, trace inactive
- trace client opens but sdev->host_offset is set to N (N > 0, from previous run)
- trace read waiting for data
- FW is resumed, tracing starts at offset 0
- trace client is woken up and it will start reading data from [N....buffer-size] -> you get stale traces read out

We can avoid this by forcing reset of sdev->host when the trace client is opened.

Reply:
@kv2019i keyon's PR is exactly about this case. We need to start copying the new trace upon resuming from N instead of 0.

Reply:
@ranj063 wrote: 
> @kv2019i keyon's PR is exactly about this case. We need to start copying the new trace upon resuming from N instead of 0.

Correct. This patch is not full replacement to Keyon's PR.

But this is worth considering. E.g. currently when you do S3 suspend, sof-logger gets disconnected. Trace client can reconnect by reopening the trace file, but it has no way to know where it should seek to when it opens. So you'll in practise get 0...64kB traces as duplicate when S3 suspend cycle occurs. This problem remains with Keyon's patch as well. With my patch, sof-logger -t output is consistent both over runtime-suspend and s3 suspend (but you have the limitations discussed above).

With Keyon's patch, you need to solve the s3 suspend issue plus implement the offset logic for BYT fw, before merging the whole thing.

OTOH, one downside is that you cannot easily use "cat" to capture the trace data during a use-case as the cat will terminate at each runtime suspend. With Keyon's patch + some fix to S3, you could avoid this.

Btw, for even more complete support, something like kernel's scheduler trace might be a better model (top/kernel/trace/trace.c and top/fs/seq_file.c) -- with seeking allowed in our interface, covering all the cases can be quite complicated but seq_file.c can help to reuse existing logic for that.

Anyways, we need to take sensible incremental steps to improve the system. This RFC is one possible such increment, or we agree to jump to offset support directly.

Reply:
@kv2019i agree with a step-by-step approach. But offset support should be added soon enough.

Reply:
@kv2019i here reset sdev->host_offset will lead to issues like this:
each time you open "sof-logger -t", the logs is empty.
That means only logs after tool run are guaranteed will be captured and output. For those logs at the first boot(user never have change to run sof-logger before that) e.g. topology parsing, pipeline creation, they themselves don't have chance to be showed to you.

Please be noticed that even end user never run "sof-logger -t", the sdev->host_offset is keep updating, so when you run the tool and reset it, you will break this updating.

Reply:
@keyonjie That is true, this is probably biggest side-effect of this RFC patch. You can workaround this by delaying module load at boot, start sof-logger and only then modprobe SOF module, but true, this does change the behaviour compared to current state.

If we don't do this, we are stuck with the problem that at open(), we have no way to synchronize the positions between client and FW and we hit more issues:
- with suspend/resume cycles, you will get repeatedly old traces in the output,
- if the FW has rolled over, you will get [0...host_offset] traces (and not the full 64KB history FW has written to the buffer

... so not running sof-logger while FW is active, is anyways problematic and without extending the buffering scheme, this is only reliable if less than 64K of traces is written before sof-logger is started.

I'll think about if we can make some special case to handle the early traces without breaking the other logic.